### ISSUE TITLE
Optax recompiles when learning rate changes

### ISSUE THREAD

**Issue by patrick-kidger** [deepmind/optax#577]
Optax isn't treating its optimisers as pytrees. This is introducing spurious recompilation:
```python
import functools as ft
import jax
import jax.numpy as jnp
import optax

optim1 = optax.adam(jnp.array(3e-3))
optim2 = optax.adam(jnp.array(3e-4))

@ft.partial(jax.jit, static_argnums=0)  # post-pytree-isation, this would probably just be `jax.jit`
def evaluate(optim, params):
    print("Compiling!")
    state = optim.init(params)
    grads = jnp.zeros_like(params)
    updates, new_state = optim.update(grads, state)
    new_params = optax.apply_updates(params, updates)
    return new_params, new_state

evaluate(optim1, jnp.array(0))  # Compiling!
evaluate(optim2, jnp.array(0))  # Compiling!
```

Can Optax optimisers be treated as pytrees with respect to their input arguments?

**Comment by patrick-kidger** [comment]
Okay, so for anyone else bumping into this issue, here's a `fixup` function that solves the issue:
```python
import types

import numpy as np
import optax
import jax.numpy as jnp
import jax.tree_util as jtu

### Library code

def _make_cell(val):
    fn = lambda: val
    return fn.__closure__[0]

def _adjust_function_closure(fn, closure):
    out = types.FunctionType(
        code=fn.__code__,
        globals=fn.__globals__,
        name=fn.__name__,
        argdefs=fn.__defaults__,
        closure=closure,
    )
    out.__annotations__.update(fn.__annotations__)
    if fn.__kwdefaults__ is not None:
        out.__kwdefaults__ = fn.__kwdefaults__.copy()
    return out

_trivial_structure = jtu.tree_structure(0)

@jtu.register_pytree_node_class
class _Closure:
    def __init__(self, fn: types.FunctionType, static: tuple[bool, ...]):
        self.fn = fn
        self.static = static

    @classmethod
    def init(cls, fn: types.FunctionType):
        cells = []
        static = []
        if fn.__closure__ is not None:
            for cell in fn.__closure__:
                if jtu.tree_structure(cell.cell_contents) == _trivial_structure and not isinstance(cell.cell_contents, (np.ndarray, jnp.ndarray)):
                    static.append(True)
                else:
                    static.append(False)
                cell = _make_cell(fixup(cell.cell_contents))
                cells.append(cell)
            fn = _adjust_function_closure(fn, tuple(cells))
        return cls(fn, tuple(static))

    def __call__(self, *args, **kwargs):
        return self.fn(*args, **kwargs)

    def tree_flatten(self):
        dynamic_cells = []
        static_cells = []
        if self.fn.__closure__ is not None:
            for cell, static in zip(self.fn.__closure__, self.static):
                if static:
                    static_cells.append(cell.cell_contents)
                    dynamic_cells.append(None)
                else:
                    static_cells.append(None)
                    dynamic_cells.append(cell.cell_contents)
        return dynamic_cells, (static_cells, self.fn)

    @classmethod
    def tree_unflatten(cls, aux, dynamic_cells):
        static_cells, fn = aux
        static = []
        if len(static_cells) > 0:
            cells = []
            for static_cell, dynamic_cell in zip(static_cells, dynamic_cells):
                if static_cell is None:
                    static.append(False)
                    cells.append(dynamic_cell)
                else:
                    static.append(True)
                    cells.append(static_cell)
            closure = tuple(_make_cell(cell) for cell in cells)
            fn = _adjust_function_closure(fn, closure)
        return _Closure(fn, tuple(static))

def _fixup_closure(leaf):
    if isinstance(leaf, types.FunctionType):
        return _Closure.init(leaf)
    else:
        return leaf

def fixup(tree):
    return jtu.tree_map(_fixup_closure, tree)

### Example usage

lr = jnp.array(1e-3)
optim = optax.chain(
    optax.adam(lr),
    optax.scale_by_schedule(
        optax.piecewise_constant_schedule(1, {200: 0.1})
    )
)
optim = fixup(optim)

# Check that we have the one leaf: the learning rate
print(jtu.tree_leaves(optim))

# Check that we can still init and update as normal.
grads = params = {"foo": jnp.array(1.)}
state = optim.init(params)
optim.update(grads, state)
```

**Comment by patrick-kidger** [comment]
(CC @cgarciae as I recall us talking about this way back when.)

**Comment by fabianp** [comment]
Closing issue given the snippet that @patrick-kidger shared solves the issue

**Comment by peabody124** [comment]
Have you guys considered incorporating this change upstream? This retracing behavior is super counterintuitive!

**Comment by colehaus** [comment]
I also encountered this recently. Note that it means that, for example, you will never hit the persistent compilation cache on your training steps if you're doing a hyperparameter sweep.

See the following example:
```py
def main(lr: ndarray[bfloat16]):
    tx = optax.adafactor(learning_rate=lr, weight_decay_rate=lr, momentum=lr, dtype_momentum=bfloat16)
    model = jnp.ones(8, bfloat16)
    opt_state = tx.init(model)
    print(jax.jit(tx.update).lower(model, opt_state, model).as_text())
```
where the learning rate gets embedded into the HLO:
```
module @jit_update_fn attributes {mhlo.num_partitions = 1 : i32, mhlo.num_replicas = 1 : i32} {
  func.func public @main(%arg0: tensor<8xbf16>, %arg1: tensor<i32>, %arg2: tensor<8xbf16>, %arg3: tensor<i32>, %arg4: tensor<8xbf16>, %arg5: tensor<8xbf16>) -> (tensor<8xbf16> {jax.result_info = "[0]"}, tensor<i32> {jax.result_info = "[1][0].count"}, tensor<1xbf16> {jax.result_info = "[1][0].v_row"}, tensor<1xbf16> {jax.result_info = "[1][0].v_col"}, tensor<8xbf16> {jax.result_info = "[1][0].v"}, tensor<i32> {jax.result_info = "[1][4].count"}, tensor<8xbf16> {jax.result_info = "[1][4].ema"}) {
    %cst = stablehlo.constant dense<4.394530e-01> : tensor<bf16>
    %cst_0 = stablehlo.constant dense<4.394530e-01> : tensor<bf16>
    ...
```

Note that it can be worked around with `inject_hyperparams`.
